File: Chapter 1/Exercise 1.2.3/8.cpp
#include <iostream>
#include <vector>
#include <bitset>
#include <math.h>
using namespace std;

int main() {
  const int N = 20;
  
  vector<int> s;
  for (int i = 0; i < N; ++i) {
    s.push_back(i);
  }
  
  vector<vector<int>> all;
  for (int i = 0; i < pow(2, N); ++i) {
    string binary = bitset<N>(i).to_string();
    vector<int> subset;
    for (int j = 0; j < binary.length(); ++j) {
      if (binary[j] == '1') {
        subset.push_back(s[j]);
      }
    }
    all.push_back(subset);
  }
  
  // print all subsets
  for (int i = 0; i < all.size(); ++i) {
    for (int j = 0; j < all[i].size(); ++j) {
      cout << all[i][j] << " ";
    }
    cout << "\n";
  }
  
  return 0;
}

File: Chapter 1/Exercise 1.2.3/8.java
import java.util.ArrayList;

class Main {
  public static void main(String[] args) {
    int N = 20;
    
    int[] s = new int[N];
    for (int i = 0; i < N; ++i) {
      s[i] = i;
    }
    
    ArrayList<ArrayList<Integer>> all = new ArrayList<>();
    for (int i = 0; i < Math.pow(2, N); ++i) {
      String binary = Integer.toBinaryString(i);
      binary = String.format("%" + N + "s", binary).replace(' ', '0');
      ArrayList<Integer> subset = new ArrayList<>();
      for (int j = 0; j < binary.length(); ++j) {
        if (binary.charAt(j) == '1') {
          subset.add(s[j]);
        }
      }
      all.add(subset);
    }
    
    // print all subsets
    for (ArrayList<Integer> subset : all) {
      System.out.println(subset);
    }
  }
}
